rules/TypeDeclaration/Rector/Property/VarAnnotationIncorrectNullableRector.php

```diff
@@ -0,0 +1,140 @@
+ */
+final class VarAnnotationIncorrectNullableRector extends AbstractRector
+{
+    private PhpDocTypeChanger $phpDocTypeChanger;
+
+    private PhpDocNullableTypeHelper $phpDocNullableTypeHelper;
+
+    private PhpDocNestedAnnotationGuard $phpDocNestedAnnotationGuard;
+
+    public function __construct(
+        PhpDocTypeChanger $phpDocTypeChanger,
+        PhpDocNullableTypeHelper $phpDocNullableTypeHelper,
+        PhpDocNestedAnnotationGuard $phpDocNestedAnnotationGuard
+    ) {
+        $this->phpDocTypeChanger = $phpDocTypeChanger;
+        $this->phpDocNullableTypeHelper = $phpDocNullableTypeHelper;
+        $this->phpDocNestedAnnotationGuard = $phpDocNestedAnnotationGuard;
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Add or remove null type from @var phpdoc typehint based on php property type declaration',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    /**
+     * @var DateTime[]
+     */
+    private ?array $dateTimes;
+}
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    /**
+     * @var DateTime[]|null
+     */
+    private ?array $dateTimes;
+}
+CODE_SAMPLE
+                ),
+
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [Property::class];
+    }
+
+    /**
+     * @param Property $node
+     */
+    public function refactor(\PhpParser\Node $node): ?\PhpParser\Node
+    {
+        if (count($node->props) !== 1) {
+            return null;
+        }
+
+        if (! $this->phpVersionProvider->isAtLeastPhpVersion(PhpVersionFeature::TYPED_PROPERTIES)) {
+            return null;
+        }
+
+        if (! $this->phpDocNestedAnnotationGuard->isPhpDocCommentCorrectlyParsed($node)) {
+            return null;
+        }
+
+        $phpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($node);
+        if (! $this->isVarDocAlreadySet($phpDocInfo)) {
+            return null;
+        }
+
+        if ($node->type === null) {
+            return null;
+        }
+
+        $phpParserType = $this->staticTypeMapper->mapPhpParserNodePHPStanType($node->type);
+
+        $varTagValueNode = $phpDocInfo->getVarTagValueNode();
+        if ($varTagValueNode === null || $varTagValueNode->type === null) {
+            return null;
+        }
+
+        $docType = $this->staticTypeMapper->mapPHPStanPhpDocTypeNodeToPHPStanType($varTagValueNode->type, $node);
+
+        $updatedPhpDocType = $this->phpDocNullableTypeHelper->resolveUpdatedPhpDocTypeFromPhpDocTypeAndPhpParserType(
+            $docType,
+            $phpParserType
+        );
+
+        if ($updatedPhpDocType === null) {
+            return null;
+        }
+
+        $this->phpDocTypeChanger->changeVarType($phpDocInfo, $updatedPhpDocType);
+
```

Comment:
`$phpDocInfo->hasChanged()` should be checked first:

```suggestion
        if (! $phpDocInfo->hasChanged()) {
            return null;
        }
```